pkg/build/builder/docker.go

```diff
@@ -416,6 +421,52 @@ func traverseAST(cmd string, node *parser.Node) int {
 	return index
 }
 
+// insertEnvAfterFrom inserts an ENV instruction with the environment variables
+// from env after every FROM instruction in the given Dockerfile instructions
+// and returns a new Dockerfile.
+func insertEnvAfterFrom(instructions []byte, env []kapi.EnvVar) ([]byte, error) {
+	if len(env) == 0 {
+		return instructions, nil
+	}
+	node, err := parser.Parse(bytes.NewReader(instructions))
+	if err != nil {
+		return nil, err
+	}
+
+	// buildEnv is an ENV Dockerfile instruction to define all environment
+	// variables from env.
+	buildEnv := []byte("ENV")
+	for _, env := range env {
+		// Marshal env.Name and env.Value as JSON strings to ensure the
+		// Dockerfile parser can read those values correctly.
+		// This covers cases when the values contain spaces or uncommon
+		// characters.
+		name, err := json.Marshal(env.Name)
+		if err != nil {
+			return nil, err
+		}
+		value, err := json.Marshal(env.Value)
+		if err != nil {
+			return nil, err
+		}
+		buildEnv = append(buildEnv, []byte(fmt.Sprintf(" %s=%s", name, value))...)
```

Comment:
I was going that way until I figured out it produces bad strings in some corner cases.
E.g.: https://github.com/docker/docker/blob/master/builder/parser/testfiles/json/Dockerfile#L8